core/src/main/java/org/apache/spark/shuffle/sort/ShuffleExternalSorter.java

```diff
@@ -353,15 +320,22 @@ private void growPointerArrayIfNecessary() throws IOException {
     assert(inMemSorter != null);
     if (!inMemSorter.hasSpaceForAnotherRecord()) {
       logger.debug("Attempting to expand sort pointer array");
-      final long oldPointerArrayMemoryUsage = inMemSorter.getMemoryUsage();
-      final long memoryToGrowPointerArray = oldPointerArrayMemoryUsage * 2;
-      final long memoryAcquired = taskMemoryManager.acquireExecutionMemory(memoryToGrowPointerArray);
-      if (memoryAcquired < memoryToGrowPointerArray) {
-        taskMemoryManager.releaseExecutionMemory(memoryAcquired);
-        spill();
-      } else {
-        inMemSorter.expandPointerArray();
-        taskMemoryManager.releaseExecutionMemory(oldPointerArrayMemoryUsage);
+      long used = inMemSorter.getMemoryUsage();
+      long needed = inMemSorter.getMemoryToExpand();
+      try {
+        acquireMemory(used + needed);  // could trigger spilling
+        if (inMemSorter.hasSpaceForAnotherRecord()) {
+          releaseMemory(used + needed);
+        } else {
+          logger.debug("Expand sort pointer array");
+          inMemSorter.expandPointerArray();
+          releaseMemory(used);
+        }
+      } catch (OutOfMemoryError oom) {
```

Comment:
Couldn't this leave the JVM in an inconsistent state? Or is the idea that `OutOfMemoryError`s thrown in this thread will perhaps be recoverable? If we're really running low on memory, I suppose there's a good chance that other threads could hit OOM as well. I guess that those threads won't have their OOMs caught and will still be able to halt the JVM.